backend/auth/auth.go

```diff
@@ -16,14 +16,25 @@ package auth
 
 import (
 	"net/http"
+	"strings"
 )
 
 type Auth interface {
-	Authenticate(next http.HandlerFunc) http.HandlerFunc
+	Authenticate(token string) (userEmail string, httpError *string)
+	Authorize(next http.HandlerFunc) http.HandlerFunc
```

Comment:
This pattern is fine for now, but remember we're eventually going to want to make authorization decisions based on what the user is actually trying to do (e.g. read list of teams, write a period etc).

When we do that, we may actually want to make the authorization decisions inside the controller methods themselves (e.g. `handlePostPeriod`).

The pattern I'd expected we would follow for that is, for the HTTP endpoints requiring authentication and authorization, we'd use the authentication middleware to determine the authenticated `User`, and pass it as an additional parameter to the controller method. Then, the controller method would make a call to the `Auth` interface to make an authorization decision.

So I thought the `Auth` interface would gain a method for each permission query that's needed, e.g.:

```
CanReadTeamList(user User) bool
CanActOnTeam(user User, team Team, action Action) bool
```

where `action` is `read`, `write` or whatever. And these methods would then be called from the controller methods, passing in the authenticated user.

We don't really want to have to replicate the parsing of the URL path that we are already doing in the HTTP mux inside the `Authorize` method.

It's fine to go this way for now for this particular PR, but you may find yourself having to refactor this (including reversing some of the changes you made here) in your next PR.

Reply:
Okay, I'll take this into account for the next PR.